### PR TITLE
Feature: Publish clip names to companion as variables

### PR DESCRIPTION
**Pull request by amarkwalder**
Hi

This feature publishes the clip names of playoutbee to companion. The names could then be used on buttons by specifying the variable in the title of the button.

In addition to mention that I was not able to test it on a version of Playoutbee 0.9.4.

Companion: 2.2.0 (2.2.0-86711173-3681) --> Experimental build
Playoutbee: 0.9.3

Regards
André

### CODE REVIEW

File: src/actions.js [hunk 1]
```diff
@@ -91,19 +91,22 @@ module.exports = {
...
 					label: 'Custom Variable',
 					id: 'variable',
-					choices:
-						this.customVariables.length === 0
-							? [{ id: 0, label: 'No variables defined' }]
-							: Object.keys(this.customVariables).map((name, index) => ({ id: index, label: name })),
+					choices: [
+						{ id: 0, label: 'No variable selected' },
+						...Object.keys(this.customVariables).map((name, index) => ({ id: (index + 1), label: name })),
+					],
```
Comment (estilles) [inline src/actions.js]: This change is completely unrelated to your PR. This allows the use of custom/user-defined variables. This has nothing to do with the internal/module variables you're proposing. Please remove.

File: src/configs.js [hunk 2]
```diff
@@ -118,6 +118,16 @@ module.exports = {
...
+			{
+				type: 'number',
+				id: 'maxClips',
+				label: 'Maximum number of clips',
+				width: 6,
+				min: 1,
+				max: 100,
+				default: 5,
+				required: true,
+			},
```
Comment (estilles) [inline src/configs.js]: The module should be able to handle however many clips are available in PlayoutBee and should dynamically create the necessary variables it wants to expose. I'm not keep on setting (or having the user set) some arbitrary limit.

File: src/feedbacks.js [hunk 3]
```diff
@@ -17,7 +17,12 @@ module.exports = {
 					type: 'dropdown',
 					label: 'Clip',
 					id: 'clip',
-					choices: this.store.clips.map(({ name }, index) => ({ id: index + 1, label: `${index + 1}: ${name}` })),
+					choices: [
+						...[...Array(this.config.maxClips).keys()].map((index) => {
+							const name = this.store.clips[index]?.name || 'Clip ' + (index + 1)
+							return { id: index + 1, label: `${index + 1}: ${name}` }
+						}),
+					],
```
Comment (estilles) [inline src/feedbacks.js]: Same as above.

File: src/variables.js [hunk 4]
```diff
@@ -92,7 +92,13 @@ module.exports = {
...
 			{ label: 'Remaining Seconds', name: 'remaining_ss' },
 			{ label: 'Remaining Fraction', name: 'remaining_ff' },
-		]
+		].concat(
+			Array(this.config.maxClips)
+				.fill( { label: undefined, name: undefined } )
+				.map( (e, i) => {
+					return { label: 'Clip ' + (i + 1), name: 'clip_' + (i + 1) }
+				})
+		)
```
Comment (estilles) [inline src/variables.js]: Variables should be dynamically created/updated based on the clips reported by PlayoutBee.

File: src/variables.js [hunk 5]
```diff
@@ -147,6 +153,9 @@ module.exports = {
 					this.store[key] = newValue
 					if (key === 'clips') {
 						this.updateVariable('clip_name', this.store.clips[this.store.clipID].name)
+						for (var i = 0; i < this.config.maxClips; i++) {
+							this.updateVariable('clip_' + (i+1), this.store.clips[i]?.name)
```
Comment (estilles) [inline src/variables.js]: I would prefer the variable names to be `clip_##_name`. There will be other clip related variables added soon (e.g. `clip_##_duration`).

Comment (estilles) [inline src/variables.js]: Nitpicking: Would prefer the use of template strings.

File: src/variables.js [hunk 6]
```diff
@@ -147,6 +153,9 @@ module.exports = {
...
+						}
```
Comment (estilles) [inline src/variables.js]: This is where I would manage the dynamic variables.

**Review by estilles — CHANGES_REQUESTED** [review]
These more or less mimic the same comments I made in your original PR.

File: src/actions.js (hunk 1 above)
Comment (amarkwalder) [reply src/actions.js]: Agree, this one slip through my commit to the branch (was not intended) and I will remove it. Nevertheless there should be a change in handling the custom variables. If you don't want to use a custom variable it will select one and the default should be not using a custom variable or am I wrong?

File: src/configs.js (hunk 2 above)
Comment (amarkwalder) [reply src/configs.js]: Intention about the config value was, that one could predefine the various buttons even if clips are not yet there. E.g. you have 3 clips within playoutbee, but you want to configure 10 buttons in companion and after you add clip 4 to playoutbee it automatically shows up in companion without doing anything. But I get your point and lets take this point up in a separate discussion, I will open an issue for that.

File: src/feedbacks.js (hunk 3 above)
Comment (amarkwalder) [reply src/feedbacks.js]: I will correct.

File: src/variables.js (hunk 4 above)
Comment (amarkwalder) [reply src/variables.js]: I will adapt.

File: src/actions.js [hunk 7]
```diff
@@ -91,22 +91,19 @@ module.exports = {
 					label: 'Clip',
 					id: 'clip',
 					choices: [
+						...this.store.clips.map(({ name }, index) => ({ id: index + 1, label: `${index + 1}: ${name}` })),
```
Comment (amarkwalder) [inline src/actions.js:94]: Reverted to original.

File: src/actions.js [hunk 8]
```diff
@@ -91,22 +91,19 @@ module.exports = {
...
 						{ id: 0, label: 'Select using custom variable' },
-						...[...Array(this.config.maxClips).keys()].map((index) => {
```
Comment (amarkwalder) [inline src/actions.js]: Reverted to original.

File: src/actions.js [hunk 9]
```diff
@@ -91,22 +91,19 @@ module.exports = {
...
-							const name = this.store.clips[index]?.name || 'Clip ' + (index + 1)
-							return { id: index + 1, label: `${index + 1}: ${name}` }
-						}),
```
Comment (amarkwalder) [inline src/actions.js]: Reverted to original.

File: src/actions.js [hunk 10]
```diff
@@ -91,22 +91,19 @@ module.exports = {
...
 					label: 'Custom Variable',
 					id: 'variable',
-					choices: [
-						{ id: 0, label: 'No variable selected' },
-						...Object.keys(this.customVariables).map((name, index) => ({ id: (index + 1), label: name })),
-					],
+					choices:
+						this.customVariables.length === 0
+							? [{ id: 0, label: 'No variables defined' }]
+							: Object.keys(this.customVariables).map((name, index) => ({ id: index, label: name })),
```
Comment (amarkwalder) [inline src/actions.js:106]: Reverted to original.

File: src/configs.js [hunk 11]
```diff
@@ -118,16 +118,6 @@ module.exports = {
...
-			{
-				type: 'number',
-				id: 'maxClips',
-				label: 'Maximum number of clips',
-				width: 6,
-				min: 1,
-				max: 100,
-				default: 5,
-				required: true,
-			},
```
Comment (amarkwalder) [inline src/configs.js]: Reverted to original.

File: src/feedbacks.js [hunk 12]
```diff
@@ -17,12 +17,7 @@ module.exports = {
 					type: 'dropdown',
 					label: 'Clip',
 					id: 'clip',
-					choices: [
-						...[...Array(this.config.maxClips).keys()].map((index) => {
-							const name = this.store.clips[index]?.name || 'Clip ' + (index + 1)
-							return { id: index + 1, label: `${index + 1}: ${name}` }
-						}),
-					],
+					choices: this.store.clips.map(({ name }, index) => ({ id: index + 1, label: `${index + 1}: ${name}` })),
```
Comment (amarkwalder) [inline src/feedbacks.js:20]: Reverted to original.

File: src/presets.js [hunk 13]
```diff
@@ -195,7 +195,7 @@ module.exports = {
 		})
 
 		// Select Clips
-		for (let index = 1; index <= this.config.maxClips; index++) {
+		for (let index = 1; index <= 50; index++) {
```
Comment (amarkwalder) [inline src/presets.js:198]: Reverted to original.

File: src/variables.js [hunk 14]
```diff
@@ -93,13 +93,13 @@ module.exports = {
 			{ label: 'Remaining Seconds', name: 'remaining_ss' },
 			{ label: 'Remaining Fraction', name: 'remaining_ff' },
 		].concat(
-			Array(this.config.maxClips)
+			Array(this.store.clips?.length || 0)
```
Comment (amarkwalder) [inline src/variables.js:96]: Used the passed clips as a base for the variables.

File: src/variables.js [hunk 15]
```diff
@@ -93,13 +93,13 @@ module.exports = {
...
 				.fill( { label: undefined, name: undefined } )
-				.map( (e, i) => {
-					return { label: 'Clip ' + (i + 1), name: 'clip_' + (i + 1) }
+				.map( (e, index) => {
+					return { label: `Clip ${index + 1} Name`, name: `clip_${index + 1}_name` }
```
Comment (amarkwalder) [inline src/variables.js:99]: Used template strings and clip_${index + 1}_name as a format

File: src/variables.js [hunk 16]
```diff
@@ -152,10 +151,9 @@ module.exports = {
 				if (JSON.stringify(oldValue) !== JSON.stringify(newValue)) {
 					this.store[key] = newValue
 					if (key === 'clips') {
+						this.initVariables()
```
Comment (amarkwalder) [inline src/variables.js:154]: Had to initialize variables as well to represent the new amount of clips passed from playoutbee

File: src/variables.js [hunk 17]
```diff
@@ -152,10 +151,9 @@ module.exports = {
...
 						this.updateVariable('clip_name', this.store.clips[this.store.clipID].name)
-						for (var i = 0; i < this.config.maxClips; i++) {
-							this.updateVariable('clip_' + (i+1), this.store.clips[i]?.name)
-						}
+						this.store.clips.forEach((clip, index) => this.updateVariable(`clip_${index + 1}_name`, clip.name))
```
Comment (amarkwalder) [inline src/variables.js:156]: update each variable with the name. Template strings are used as well according to the previous comment.

**Comment by istnv** [comment]
VLC has a configuration option to specify how many clip name variables to reserve. They preset to blank or '·'. This enables you to make a page with 10 `Play #` buttons that will show the actual clip name from VLC or blank instead of `$NA` if there are fewer clips.
